### ISSUE TITLE
Net::IP creates wrong IPv4 reverse name when last octet is "0"

### ISSUE THREAD

**Issue by matsduf** [zonemaster/zonemaster-engine#1107]
Net::IP::XS is used by Zonemaster i.e. to create reverse names of IP addresses. If Net::IP::XS is not available, Net::IP is used instead. The issue reported here is affecting Net::IP and not Net::IP::SX, which means that OSs that do load Net::IP::SX are not affected (FreeBSD, Debian, Ubuntu).

Affected OSs are CentOS 7 Linux and Docker. It is unknown if Rocky Linux is affected.

A work around for affected OSs can be, if possible, to install Net::IP::XS. Version 0.21 works with Zonemaster. If lower version also works has not been tested.

## Issue
Creating the reverse name using Net::IP when the last octet **is not** "0":
```
$ perl -E 'use Net::IP; $ip = new Net::IP ("10.11.12.13"); say $ip->reverse_ip'
13.12.11.10.in-addr.arpa
```
Creating the reverse name using Net::IP when the last octet **is** "0":
```
$ perl -E 'use Net::IP; $ip = new Net::IP ("10.11.12.0"); say $ip->reverse_ip'
12.11.10.in-addr.arpa.
```
*Note that the first label, that corresponds to the last octet, is missing.*

The same thing using Net::IP::XS:
```
$ perl -E 'use Net::IP::XS; $ip = new Net::IP::XS ("10.11.12.13"); say $ip->reverse_ip'
13.12.11.10.in-addr.arpa.

$ perl -E 'use Net::IP::XS; $ip = new Net::IP::XS ("10.11.12.0"); say $ip->reverse_ip'
0.12.11.10.in-addr.arpa.
```
*Note that both reverse names are correct.*

## IPv6 reverse
IPv6 reverse is not affected by the issue.
```
$ perl -E 'use Net::IP; $ip = new Net::IP ("2a05:d016:c84:7600::1"); say $ip->reverse_ip'
1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.7.4.8.c.0.6.1.0.d.5.0.a.2.ip6.arpa.

$ perl -E 'use Net::IP; $ip = new Net::IP ("2a05:d016:c84:7600::0"); say $ip->reverse_ip'
0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.7.4.8.c.0.6.1.0.d.5.0.a.2.ip6.arpa.
```

## Ubuntu

When Ubuntu 22.04 is installed using the CPAN method, Net::IP::XS is not installed, but it is till not affected. Net::IP is claimed to be the same version (1.26) but it has been updated in the relevant part of the code. Correct output on Ubuntu:
```
$ perl -E 'use Net::IP; $ip = new Net::IP ("10.11.12.0"); say $ip->reverse_ip'
0.12.11.10.in-addr.arpa.
```

## Reference

This issue was discovered and reported by @anandb-ripencc in #1105.

The bug in Net::IP is reported in https://rt.cpan.org/Public/Bug/Display.html?id=129964

**Comment by matsduf** [comment]
Resolved by #1108

**Comment by anandb-ripencc** [comment]
Can this be fixed just by installing Net::IP::XS?

**Comment by matsduf** [comment]
If version 0.21 is available is should work, maybe with lower version, but I have not tested. That is the version of Net::IP::XS installed on FreeBSD, where it works. The version of Net::IP is 1.26 on FreeBSD (and CentOS 7).

Nothing has to be changed in Zonemaster to use Net::IP::XS. If it is installed it will be used.

**Comment by anandb-ripencc** [comment]
@matsduf I resolved it in our build of Zonemaster, by explicitly installing Net:IP::XS. The build process pulled in version 0.21. With this, #1108 is not needed, if you adjust Zonemaster's dependencies and specify Net:IP::XS explicitly.

**Comment by matsduf** [comment]
@anandb-ripencc, I am glad that you could resolve the issue so easily. I do not know why Net::IP::XS is not included as a dependency for CentOS Linux 7.

The fix in #1018 is, however, still needed. Firstly, we know that the Docker installation is affected. It is unknown if the Rocky Linux is affected. Secondly, as long as Zonemaster supports running with Net::IP, which is an alternative dependency, the bug must be worked around.

**Comment by anandb-ripencc** [comment]
I just checked the installation Instructions for CentOS 7 and Rocky Linux. Two things stand out:
1. You need to add a version number to the Rocky Linux instructions, because now we have both Rocky 8 and 9. The current instructions may be correct for Rocky 8, but not for Rocky 9.
2. None of the instructions ask for installation of Net::IP::XS (whereas they do for FreeBSD). So you need to update the installation document and add this module, since Rocky 8 is also affected.

Finally, if you just depend on Net::IP::XS everywhere, you can just drop the dependency on Net::IP, so you don't need to work around any bug. I feel this is the better solution.

**Comment by matsduf** [comment]
Thank you, good observations regarding Rocky Linux. And thank you for investigating.

Yes, it is version 8 (8.6) that Zonmaster supports. In the installation instructions it does not say which version of the OS that is supported (except for CentOS 7). Instead it says 

> For details on supported operating system versions and Perl verisons for
> Zonemaster::Engine, see the [declaration of prerequisites].

And there it says what OS version that is supported. Maybe that is not clear enough.

Rocky Linux 9 was released after the release of the latest Zonemaster release.

There is a reason why Zonemaster supports both Net::IP::XS and Net::IP and that is that the former has not been available for all OSs. Today it might be available for all OSs, but I do not think it is available for Alpine Linux used for the Docker image.

Please create an issue on the lack of Net::IP::XS in the installation instructions for CentOS 7 and Rocky Linux. For 

Currently we have three options for the bug (since there is little hope that the Net::IP maintainer should do any updates now after years of inactivity):

1. Do nothing.
2. Update the code to remove the support for Net::IP with the risk that some platform does not support Zonemaster anymore, e.g. Docker.
3. Update the code with a work-around.

I chose number 3, but when approach the release we might decide to go for number 2 (assuming that all platforms has support for Net::IP::XS). There could possibly be alternatives for Alpine linux for the Docker image.

[Declaration of prerequisites]:                      https://github.com/zonemaster/zonemaster#prerequisites

**Comment by anandb-ripencc** [comment]
Hi Mats,

As I wrote in another thread, Net::IP::XS can be installed everywhere, from CPAN, using cpanm. Therefore, the issue of whether it is available as a package in some distro or another is meaningless.

**Comment by matsduf** [comment]
This is just a description of the current situation. Nothing else.

**Comment by matsduf** [comment]
Resolved by #1119
